_episodes/02-keras.md

```diff
@@ -0,0 +1,514 @@
+> determine the relavent inputs and outputs.
+{:.keypoints}
+## 3. Prepare data
+The input data and target data are not yet in a format that is suitable to use for training a neural network.
+During the exploration phase you may have noticed that some rows in the dataset have missing (NaN)
+values, leaving such values in the input data will ruin the training, so we need to deal with them.
+There are many ways to deal with missing values, but for now we will just remove the offending rows.
+
+Add a call to `dropna()` before the input_data definition that we used above, the cell should now
+look like this:
+~~~
+valid_data = penguins.dropna()
+
+training_data = valid_data.drop(columns=["species", 'island', 'sex'])
+target = valid_data['species']
+~~~
+{:.language-python}
+
+Second, the target data is also in a format that cannot be used to train.
+To train a neural network we need to be able to calculate how "far away" the species predicted by the
+neural network is from the true species.
+When the target is a string category column as we have here it is very difficult to determine this "distance" or error.
+Therefore we will transform this column into a more suitable format.
+Again there are many ways to do this, however we will be using the 1-hot encoding.
+This encoding creates multiple columns, as many as there are unique values, and
+puts a 1 in the column with the corresponding correct class, and 0's in
+the other columns.
+For instance, for a penguin of the Adelie species the 1 hot encoding would be 1 0 0
+
+Fortunately pandas is able to generate this encoding for us.
+~~~
+target = pd.get_dummies(valid_data['species'])
+target.head() # print out the top 5 to see what it looks like.
+~~~
+{:.language-python}
+
+> ## Keras for Neural Networks
+> For this lesson we will be using [Keras](https://keras.io/) to define and train our neural network
+> models.
+> Keras is a machine learning framework with ease of use as one of its main features.
+> It is part of the tensorflow python package and can be imported using `from tensorflow import keras`.
+>
+> Keras includes functions, classes and definitions to define deep learning models, cost functions and
+> optimizers (optimizers are used to train a model).
+{:.callout}
+
+## 4. Choose a cost function and metrics
+Having selected the target enoding, we need to select an appropriate loss
+function that we will use during training.
+This loss function tells the training algorithm how wrong, or how 'far away' from the true
+value the predicted value is.
+
+For the one hot encoding we will use the Categorical Crossentropy loss.
+This is a measure for how close the distribution of the three neural network outputs corresponds
+to the distribution of the three values in the one hot encoding.
+It is lower if the distributions are more similar.
+In keras this is implemented in the `keras.losses.CategoricalCrossentropy` class.
```

Comment:
Refer to the documentation of Keras here? To show that there's different types of loss for different types of problems.

Reply:
You could also add here that you would want additional metrics to measure your performance such as accuracy. Or keep that for the next episode?